persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/binary/JdbcBinaryStore.java

```diff
@@ -357,16 +308,111 @@ public void purge(Executor threadPool, PurgeListener task) {
+         connection.setAutoCommit(false);
+
+         // We load all existing buckets up front to prevent multiple SQL statements loading/writing the same bucket
+         Map<Integer, Bucket> existingBuckets = getExistingBuckets(connection, batchModification);
+         Set<Bucket> newBuckets = updateAndCreateBuckets(batchModification.getMarshalledEntries(),
+                                                         batchModification.getKeysToRemove(), existingBuckets);
+
+         // Write changes to DB
+         try (PreparedStatement insertBatch = connection.prepareStatement(tableManager.getInsertRowSql());
+              PreparedStatement updateBatch = connection.prepareStatement(tableManager.getUpdateRowSql())) {
+
+            for (Bucket bucket : existingBuckets.values()) {
+               if (newBuckets.contains(bucket)) {
+                  prepareWriteStatement(insertBatch, bucket, tableManager.getInsertRowSql());
+                  insertBatch.addBatch();
+               } else {
+                  prepareWriteStatement(updateBatch, bucket, tableManager.getUpdateRowSql());
+                  updateBatch.addBatch();
+               }
+            }
+            insertBatch.executeBatch();
+            updateBatch.executeBatch();
+         }
+      } catch (SQLException | InterruptedException e) {
+         throw log.prepareTxFailure(e);
+      }
+   }
+
+   private Map<Integer, Bucket> getExistingBuckets(Connection connection, BatchModification batchModification) throws SQLException {
+      Set<Integer> bucketIds = batchModification.getAffectedKeys().stream()
+            .map(this::getBuckedId)
+            .collect(Collectors.toSet());
+
+      String sql = tableManager.getSelectMultipleRowSql(bucketIds.size());
+      if (sql == null)
+         return new HashMap<>();
+
+      try (PreparedStatement ps = connection.prepareStatement(sql)) {
+         int count = 0;
+         for (Integer id : bucketIds)
+            ps.setInt(++count, id);
+
+         try (ResultSet rs = ps.executeQuery()) {
+            Map<Integer, Bucket> existingBuckets = new HashMap<>();
+            while (rs.next()) {
+               Bucket bucket = loadBucket(rs);
+               existingBuckets.put(bucket.getBucketId(), bucket);
+            }
+            return existingBuckets;
+         }
+      }
+   }
+
+   // Returns a set of newly created Buckets
+   private Set<Bucket> updateAndCreateBuckets(Collection<MarshalledEntry> modifiedEntries, Set<Object> keysToDelete,
+                                              Map<Integer, Bucket> existingBuckets) {
+      Set<Bucket> newBuckets = new HashSet<>();
+      for (MarshalledEntry entry : modifiedEntries) {
+         Integer bucketKey = getBuckedId(entry.getKey());
+         Object entryKey = entry.getKey();
+         InternalMetadata m = entry.getMetadata();
+         Bucket existingBucket = existingBuckets.get(bucketKey);
+
+         // How should we handle expired entries in a Tx?
```

Comment:
Good spot! This should have been removed. The logic is the same as when we write for an expired entry outside of a tx, i.e. we remove the entry from the bucket if it has expired and continue to execute the rest of the Tx. I'll remove the comment and re-push